src/meta/src/barrier/mod.rs

```diff
@@ -162,6 +167,131 @@ pub struct GlobalBarrierManager<S: MetaStore> {
     metrics: Arc<MetaMetrics>,
 
     env: MetaSrvEnv<S>,
+
+    /// The max barrier nums in flight
+    in_flight_barrier_nums: usize,
+}
+
+struct CheckpointControl<S: MetaStore> {
+    /// Signal whether to start recovery
+    is_recovery: bool,
+    /// Signal whether the barrier with building actor is sent
+    is_build_actor: bool,
+    /// Save the state and message of barrier in order
+    command_ctx_queue: VecDeque<EpochNode<S>>,
+}
+
+impl<S> CheckpointControl<S>
+where
+    S: MetaStore,
+{
+    fn new() -> Self {
+        Self {
+            is_recovery: false,
+            is_build_actor: false,
+            command_ctx_queue: VecDeque::default(),
+        }
+    }
+
+    /// Return the nums of barrier (the nums of in-flight-barrier , the nums of all-barrier)
+    fn get_barrier_len(&self) -> (usize, usize) {
+        (
+            self.command_ctx_queue
+                .iter()
+                .filter(|x| matches!(x.state, InFlight))
+                .count(),
+            self.command_ctx_queue.len(),
+        )
+    }
+
+    fn finish_recovery(&mut self) {
+        self.is_recovery = false;
+    }
+
+    /// Inject a `command_ctx` in `command_ctx_queue`, and it's state is `InFlight`.
+    fn inject(
+        &mut self,
+        command_ctx: Arc<CommandContext<S>>,
+        notifiers: SmallVec<[Notifier; 1]>,
+        timer: HistogramTimer,
+    ) {
+        self.command_ctx_queue.push_back(EpochNode {
+            timer: Some(timer),
+            result: None,
+            state: InFlight,
+            command_ctx: command_ctx.clone(),
+            notifiers,
+        });
+        if !matches!(command_ctx.command, Command::Plain(_)) {
+            self.is_build_actor = true;
+        }
+    }
+
+    /// Change the state of this `prev_epoch` to `Complete`. Return nodes from the first node to the
+    /// first node with `InFlight` [`Complete`..`InFlight`) and remove them.
+    fn complete(
+        &mut self,
+        prev_epoch: u64,
+        result: Result<Vec<BarrierCompleteResponse>>,
+    ) -> VecDeque<EpochNode<S>> {
+        if let Some(node) = self
+            .command_ctx_queue
+            .iter_mut()
+            .find(|x| x.command_ctx.prev_epoch.0 == prev_epoch)
+        {
+            assert!(matches!(node.state, InFlight));
+            node.state = Complete;
+            node.result = Some(result);
+        };
+        let index = match self.command_ctx_queue.iter().find_position(|x| {
+            if !matches!(x.command_ctx.command, Command::Plain(_)) && !matches!(x.state, InFlight) {
+                self.is_build_actor = false;
+            };
+            matches!(x.state, InFlight)
```

Comment:
I suggest we directly check Complete state instead of checking the opposite InFlight state to make the code easier to understand and less prone to bugs in case we need to add new states in the future.

Reply:
I want to find continuous `Complete` state , so I find the first `InFilght`.  I think i should modify it